logstash-core/lib/logstash/runner.rb

```diff
@@ -601,4 +604,19 @@ def setting(key)
     @settings.get_value(key)
   end
 
+  def configure_pipeline_buffer_type
+    # if user hasn't explicitly set Netty's interested property
+    if java.lang.System.getProperty("io.netty.noPreferDirect") == nil
+      if setting("pipeline.buffer.type") == "heap"
+        # set Netty's Java properties to prefer heap memory allocation
+        java.lang.System.setProperty("io.netty.noPreferDirect", "true")
+      else
+        # set Netty's Java properties to prefer direct memory allocation
+        java.lang.System.setProperty("io.netty.noPreferDirect", "false")
+      end
+    else
+      logger.warn("Can't switch IO buffer to Java heap allocation because 'io.netty.noPreferDirect' Java property was already set in jvm.options file.")
```

Comment:
We also execute this code for pipeline.buffer.type == direct, so this should be changed (also we don't know if the noPreferDirect java property was set via jvm.options or LS_JAVA_OPTS):

```suggestion
      logger.warn("Ignoring 'pipeline.buffer.type' since the 'io.netty.noPreferDirect' Java property has already been set (check LS_JAVA_OPTS or jvm.options file.")
```